### ISSUE TITLE
Add to stage event

### ISSUE THREAD

**Issue by peteshand** [HeapsIO/heaps#718]
Hi Nicolas

I'm looking at adding support for Heaps within Robotlegs.
However to get mediator mapping working I really need a way of listening for on add/remove of display objects from the root display object (Stage, Window, etc). Looking at the source code of heaps there doesn't appear to be this type of functionally. Would you be open to a pull request adding this?

**Comment by Yanrishatum** [comment]
There are events like that - `onAdd`, `onRemove` and `onHierarchyMoved`.
`onAdd` gets called when object is allocated, which means that it's added to allocated scene. Where `allocated` means it's an active scene. There are other cases where it can be called however - when using `drawTo` for example.
`onRemove` gets called when object is removed from scene or scene is disposed (e.g. switching active scene). But keep in mind that if you switch the scene without disposing flag you won't get `onRemove` call.
Lastly, `onHierarchyMoved` called whenever object is moved within allocated scene. This is a separate event in order to avoid double allocation when user wants to move object within hierarchy.

In short, look better.

**Comment by peteshand** [comment]
Hey @Yanrishatum 
Thanks for taking the time to reply.

Yeah I did see onAdd, onRemove, but I couldn't see how they propagate to the top level display object (Window in the case of Heaps).

What I'm looking for is something to the effect of:
```
Window.inst.onChildAdded((child:h3d.scene.Object) -> {
   // Do something
);

Window.inst.onChildRemoved((child:h3d.scene.Object) -> {
   // Do something
);
```

**Comment by Yanrishatum** [comment]
Okay, there is no such event system. And I don't see much point in it apart from some weird hooks trying to scaffold other engine on top. Again, I, personally, don't see good reasons to add such propagation system. 
Additionally I don't think adding such events to Window is a good idea. First, Heaps not just 3D engine, it contains 2 different scenes - 2D and 3D, and they are incompatible with each other. At best you want to add those to Scene instance, but even here it's still questionable addition. Even more, Heaps isn't specifically hardlocked on just 2 types of Scenes, you can create and use your own Scene with custom renderer and such, completely ignoring stock Scenes (for whatever reason).
I'd say just use either macro to add hook like that in your integration or just use custom Object that would propagate it's addition. For reference, here's my macro extensions for Heaps: https://github.com/Yanrishatum/heeps/blob/master/hxd/heeps/Extensions.hx

To reiterate - that is a very specific feature you're looking for, and it is not required for majority of people, and better off being implemented on the side of framework you trying to attach. I'm not an official Heaps dev, so maybe @ncannasse would agree with you, but that's my personal position - questionable addition. ¯\\\_(ツ)_/¯

**Comment by ncannasse** [comment]
I agree with @Yanrishatum, the event is not required for Heaps to run and adding it would create non zero overhead, so better having it as a custom feature than directly in the base framework

**Comment by peteshand** [comment]
I have it working locally... it adds zero additional overhead if listeners aren't added.. (which only happen if you're utilizing the feature) and even then its going to be negligible. It's only about 20 lines of code. If you're interested in reviewing let me know.

A macro to patch heaps externally is a nice idea though, so I'll try find some time to look into that.
